File: backend/src/main/java/com/upblit/backend/query/model/Trace.java
package com.upblit.backend.query.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;

import java.time.Instant;
import java.util.Map;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "traces")
public class Trace {

    @Id
    private String id;

    @Field("clientTimestamp")
    private Instant clientTimestamp;

    @Field("serverTimestamp")
    private Instant serverTimestamp;

    @Field("projectId")
    private Long projectId;

    @Field("applicationId")
    private Long applicationId;

    @Field("requestMethod")
    private String requestMethod;

    @Field("requestURL")
    private String requestURL;

    @Field("responseStatus")
    private Integer responseStatus;

    @Field("traceId")
    private String traceId;

    @Field("spanType")
    private String spanType;

    @Field("spanId")
    private String spanId;

    @Field("parentSpanId")
    private String parentSpanId;

    @Field("durationMs")
    private Long durationMs;

    @Field("timestamp")
    private Instant timestamp;

    @Field("meta")
    private Map<String, String> meta;
}

File: ingestor/models/Trace.go
package models

import (
	"time"
)

type Trace struct {
	ClientTimestamp time.Time `json:"clientTimestamp" bson:"clientTimestamp"`
	ServerTimestamp time.Time `json:"serverTimestamp" bson:"serverTimestamp"`

	ProjectID     int64 `json:"projectId" bson:"projectId"`
	ApplicationID int64 `json:"applicationId" bson:"applicationId"`

	RequestMethod  string            `json:"requestMethod" bson:"requestMethod"`
	RequestURL     string            `json:"requestURL" bson:"requestURL"`
	ResponseStatus int               `json:"responseStatus" bson:"responseStatus"`
	TraceID        string            `json:"traceId" bson:"traceId"`
	SpanType       string            `json:"spanType" bson:"spanType"`
	SpanID         string            `json:"spanId" bson:"spanId"`
	ParentSpanID   string            `json:"parentSpanId" bson:"parentSpanId"`
	DurationMs     int64             `json:"durationMs" bson:"durationMs"`
	Timestamp      time.Time         `json:"timestamp" bson:"timestamp"`
	Meta           map[string]string `bson:"meta" json:"meta"`
}
